openwisp_ipam/base/models.py

```diff
@@ -82,20 +84,35 @@ def _validate_multitenant_master_subnet(self):
             return
         if self.master_subnet.organization:
             self._validate_org_relation('master_subnet', field_error='master_subnet')
-        elif self.organization:
+
+    def _validate_multitenant_unique_child_subnet(self):
+        if self.master_subnet is None or self.master_subnet.organization_id is not None:
+            return
+        qs = self._meta.model.objects.exclude(id=self.pk).filter(subnet=self.subnet)
+        if qs.exists():
             raise ValidationError(
                 {
-                    'master_subnet': _(
-                        'Please ensure that the organization of this subnet and '
-                        'the organization of the related subnet match.'
+                    'subnet': _(
+                        'This subnet is already assigned to another organization.'
                     )
                 }
             )
 
     def _validate_overlapping_subnets(self):
-        qs = self._meta.model.objects.filter(organization=self.organization).only(
-            'subnet'
-        )
+        organization_query = Q(organization_id=self.organization_id)
+        error_message = _('Subnet overlaps with {0}.')
+        if self.master_subnet:
+            if self.master_subnet.organization_id is None:
```

Comment:
this works if there's only 1 level, but may not work if the master subnet does have an org set but has itself a relation to another master subnet which is shared, that's why in one of the validation checks below we iterate over all the master subnets